app/src/main/java/org/mozilla/fenix/components/metrics/AppStartupTelemetry.kt

```diff
@@ -0,0 +1,159 @@
+import androidx.lifecycle.Lifecycle
+import androidx.lifecycle.LifecycleObserver
+import androidx.lifecycle.OnLifecycleEvent
+import androidx.lifecycle.ProcessLifecycleOwner
+import mozilla.components.support.utils.SafeIntent
+import org.mozilla.fenix.components.metrics.Event.AppAllStartup
+import org.mozilla.fenix.components.metrics.Event.AppAllStartup.Source
+import org.mozilla.fenix.components.metrics.Event.AppAllStartup.Source.APP_ICON
+import org.mozilla.fenix.components.metrics.Event.AppAllStartup.Source.ERROR
+import org.mozilla.fenix.components.metrics.Event.AppAllStartup.Source.CUSTOM_TAB
+import org.mozilla.fenix.components.metrics.Event.AppAllStartup.Source.LINK
+import org.mozilla.fenix.components.metrics.Event.AppAllStartup.Source.UNKNOWN
+
+import org.mozilla.fenix.components.metrics.Event.AppAllStartup.Type
+import org.mozilla.fenix.components.metrics.Event.AppAllStartup.Type.COLD
+import org.mozilla.fenix.components.metrics.Event.AppAllStartup.Type.WARM
+import org.mozilla.fenix.components.metrics.Event.AppAllStartup.Type.HOT
+
+/**
+ * Tracks application startup source, type, and whether or not activity has savedInstance to restore
+ * the activity from. Sample metric = [source = COLD, type = APP_ICON, hasSavedInstance = false]
+ * The basic idea is to collect these metrics from different phases of startup through
+ * [AppAllStartup] and finally report them on Activity's onResume() function.
+ */
+@Suppress("TooManyFunctions")
+class AppStartupTelemetry(private val metrics: MetricController) : LifecycleObserver {
+
+    init {
+        ProcessLifecycleOwner.get().lifecycle.addObserver(this)
+    }
+
+    private var isMetricRecordedSinceAppWasForegrounded = false
+    private var wasAppCreateCalledBeforeActivityCreate = false
+
+    private var onCreateData: AppAllStartup? = null
+    private var onRestartData: Pair<Type, Boolean>? = null
+    private var onNewIntentData: Source? = null
+
+    fun onFenixApplicationOnCreate() {
+        wasAppCreateCalledBeforeActivityCreate = true
+    }
+
+    fun onHomeActivityOnCreate(safeIntent: SafeIntent, hasSavedInstanceState: Boolean) {
+        onCreateData = AppAllStartup(
+            getStartupSourceFromIntent(safeIntent, false),
+            getAppStartupType(),
+            hasSavedInstanceState
+        )
+        wasAppCreateCalledBeforeActivityCreate = false
+    }
+
+    fun onExternalAppBrowserOnCreate(safeIntent: SafeIntent, hasSavedInstanceState: Boolean) {
+        onCreateData = AppAllStartup(
+            getStartupSourceFromIntent(safeIntent, true),
+            getAppStartupType(),
+            hasSavedInstanceState
+        )
+        wasAppCreateCalledBeforeActivityCreate = false
+    }
+
+    /**
+     * we are not setting [Source] in this method since source is derived from a [SafeIntent].
+     * therefore source either gets set in oncreate() or onNewIntent().
+     */
+    fun onHomeActivityOnRestart() {
+        onRestartData = Pair(HOT, false)
```

Comment:
I just noticed: since it's always `false`, can't we just omit this extra in the telemetry? (e.g. a nullable field in the metric class that, when null, does not add the extra)

Reply:
not sure what you mean by omit extra in telemetry?

Reply:
Instead of recording [HOT, APP_ICON, false], record [HOT, APP_ICON]

Reply:
Be sure to document the fact that it's optional